src/wallet/sqlite.cpp

```diff
@@ -251,22 +251,106 @@ void SQLiteBatch::Close()
 
 bool SQLiteBatch::ReadKey(CDataStream&& key, CDataStream& value)
 {
-    return false;
+    if (!m_database.m_db) return false;
+    assert(m_database.m_read_stmt);
+
+    // Bind: leftmost parameter in statement is index 1
+    int res = sqlite3_bind_blob(m_database.m_read_stmt, 1, key.data(), key.size(), SQLITE_STATIC);
```

Comment:
I think we've discussed it already here https://github.com/bitcoin/bitcoin/pull/19077#discussion_r493400426

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19077#discussion_r500018582

> I think we've discussed it already here [#19077 (comment)](https://github.com/bitcoin/bitcoin/pull/19077#discussion_r493400426)

You can use the same prepared statement from different threads but you can't use the same statement for different purposes with different values at the same time from different threads. The only thing that might prevent bugs in practice is the fact that most of the database is copied into memory and not actually accessed through database lookups, and the fact that the cs_wallet mutex makes most wallet operations single threaded in practice. But these things only make the code accidentally correct instead of correct by design. The sqlite implementation shouldn't have object sharing and reference counting code that makes it more complicated than it needs to be, and less safe than the bdb implementation (or any sane key/value API). You shouldn't get undefined results, for example, simply reading two keys from two different threads.

Reply:
Now I think you're correct. Looks like the locking in sqlite will guarantee safety for a single call, but we ourselves still have to take care for the sequence of the calls.

> You can use the same prepared statement from different threads but you can't use the same statement for different purposes with different values at the same time from different threads.

This is what I failed to articulate before and was to quick to agree it's safe.

Reply:
Moved all of the statements into `SQLiteBatch`.